orttraining/orttraining/core/session/tensor_helper.cc

```diff
@@ -0,0 +1,231 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/core/session/tensor_helper.h"
+#include "core/providers/cuda/cuda_common.h"
+
+namespace onnxruntime {
+namespace training {
+
+// Return the shape of a tensor slice.
+std::vector<int64_t> GetSliceShape(
+    const std::vector<int64_t>& shape,  // before-slicing tensor shape
+    const size_t& slice_axis,           // axis to slice along
+    const size_t& num_slices) {         // number of slices along the slicing axis
+  ORT_ENFORCE(shape.size() > 0);
+  ORT_ENFORCE(slice_axis < shape.size());
+  ORT_ENFORCE(num_slices > 0);
+
+  // Shape of slice along slice_axis.
+  std::vector<int64_t> slice_shape;
+
+  // Go through the original dimensions to get the dimensions after slicing.
+  for (size_t i_shape = 0; i_shape < shape.size(); ++i_shape) {
+    const auto d = shape[i_shape];
+
+    // Compute slice's shape.
+    if (i_shape == slice_axis) {
+      // Dimension shrinks due to slicing.
```

Comment:
>// Dimension shrinks due to slicing. [](start = 6, length = 36)

It may worth a check to ensure d can be divided by num_slices

Reply:
Ok.

---
In reply to: [538746193](https://github.com/microsoft/onnxruntime/pull/5815#discussion_r538746193) [](ancestors = 538746193)